File: skills/playwright/src/browser.rs
use chromiumoxide::browser::{Browser, BrowserConfig};
use chromiumoxide::Page;
use futures::StreamExt;

pub struct BrowserSession {
    pub page: Page,
}

impl BrowserSession {
    pub async fn url(&self) -> String {
        self.page.url().await.ok().flatten().unwrap_or_default().to_string()
    }

    pub async fn title(&self) -> String {
        self.page
            .evaluate("document.title")
            .await
            .ok()
            .and_then(|v| v.into_value::<String>().ok())
            .unwrap_or_default()
    }
}

pub async fn get_browser() -> anyhow::Result<BrowserSession> {
    // For CLI usage, create a fresh browser each time
    let config = BrowserConfig::builder()
        .no_sandbox()
        .window_size(1280, 720)
        .arg("--disable-setuid-sandbox")
        .arg("--headless=new")
        .build()
        .map_err(|e| anyhow::anyhow!("Browser config error: {e}"))?;

    let (browser, mut handler) = Browser::launch(config).await?;

    // Spawn handler task
    tokio::spawn(async move {
        while let Some(event) = handler.next().await {
            // Process browser events
            let _ = event;
        }
    });

    let page = browser.new_page("about:blank").await?;

    // Store for later cleanup
    // Note: for the CLI, each invocation is a new process, so we don't need
    // to worry about browser reuse across commands
    Ok(BrowserSession { page })
}

pub async fn close_browser() {
    // Browser will be dropped when the process exits
    // For explicit cleanup, we could store the browser handle
}

pub async fn get_status() -> serde_json::Value {
    serde_json::json!({
        "connected": false,
        "note": "Each CLI invocation creates a fresh browser"
    })
}

File: skills/playwright/src/browser.ts
import { homedir } from "node:os";
import { join } from "node:path";
import {
	type Browser,
	type BrowserContext,
	chromium,
	type Page,
} from "playwright";

const SESSION_DIR = join(homedir(), ".mini-claw", "playwright");

export interface BrowserSession {
	browser: Browser;
	context: BrowserContext;
	page: Page;
}

// Module-level browser instance for reuse within same process
let browserInstance: Browser | null = null;
let contextInstance: BrowserContext | null = null;
let pageInstance: Page | null = null;

async function launchBrowser(): Promise<Browser> {
	const browser = await chromium.launch({
		headless: true,
		args: ["--no-sandbox", "--disable-setuid-sandbox"],
	});
	return browser;
}

export async function getBrowser(): Promise<BrowserSession> {
	// Reuse existing browser if available
	if (browserInstance && browserInstance.isConnected()) {
		if (contextInstance && pageInstance) {
			return {
				browser: browserInstance,
				context: contextInstance,
				page: pageInstance,
			};
		}
	}

	// Launch new browser
	browserInstance = await launchBrowser();

	contextInstance = await browserInstance.newContext({
		viewport: { width: 1280, height: 720 },
		userAgent:
			"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
	});

	pageInstance = await contextInstance.newPage();

	return {
		browser: browserInstance,
		context: contextInstance,
		page: pageInstance,
	};
}

export async function closeBrowser(): Promise<void> {
	if (browserInstance) {
		await browserInstance.close();
		browserInstance = null;
		contextInstance = null;
		pageInstance = null;
	}
}

export async function getStatus(): Promise<{
	connected: boolean;
	url?: string;
	title?: string;
}> {
	if (!browserInstance || !browserInstance.isConnected()) {
		return { connected: false };
	}

	try {
		if (pageInstance) {
			const url = pageInstance.url();
			const title = await pageInstance.title();
			return {
				connected: true,
				url: url === "about:blank" ? undefined : url,
				title: title || undefined,
			};
		}
		return { connected: true };
	} catch {
		return { connected: false };
	}
}

export function getSessionDir(): string {
	return SESSION_DIR;
}
